### PR TITLE
[Communication]: Create test coverage report in sdk pipelines

### PR DESCRIPTION
**Pull request by JianpingChen**

### CODE REVIEW

File: eng/pipelines/templates/stages/archetype-sdk-tests.yml [hunk 1]
```diff
@@ -114,6 +114,8 @@ stages:
   - ${{ if or(contains(parameters.Clouds, cloud.key), and(contains(variables['Build.DefinitionName'], 'tests-weekly'), contains(parameters.SupportedClouds, cloud.key))) }}:
     - ${{ if not(contains(parameters.UnsupportedClouds, cloud.key)) }}:
       - stage: ${{ cloud.key }}_${{ parameters.TestName }}
+        variables:
```
Comment (minnieliu) [inline eng/pipelines/templates/stages/archetype-sdk-tests.yml:117]: Why do we need to add this variable? Since this yml file is shared between all the SDKs,  can you ran another SDK's pipeline to see if it will cause any regression?

Comment (JianpingChen) [reply eng/pipelines/templates/stages/archetype-sdk-tests.yml:117]: Introducing a variable no one else is using feels pretty safe to me. :-) 

Sure, let me try one just to be sure.

Comment (sacheu) [reply eng/pipelines/templates/stages/archetype-sdk-tests.yml:117]: Why we need this new variable?

**Review by minnieliu — APPROVED** [review]

File: eng/pipelines/templates/stages/archetype-sdk-tests.yml (hunk 1 above)
Comment (sacheu) [reply eng/pipelines/templates/stages/archetype-sdk-tests.yml:117]: I see that we only publish coverage only in Public cloud.

Comment (benbp) [reply eng/pipelines/templates/stages/archetype-sdk-tests.yml:117]: This value should already be available via `$(System.StageName)`.

Comment (benbp) [reply eng/pipelines/templates/stages/archetype-sdk-tests.yml:117]: I'm open to a more clearly named variable for this purpose, but to @minnieliu's point, it will need to be done in a way that is common to not only all pipelines, but all languages. That would likely mean doing it here: https://github.com/Azure/azure-sdk-tools/blob/master/eng/common/pipelines/templates/jobs/archetype-sdk-tests-generate.yml#L92

It's a little tricky because of how you're trying to reference something at this top level to be called lazily, meaning we can't reference `${{ parameters.CloudConfig.Cloud }}` which would be accessible at the job yaml definition scope.

My thinking right now is let's keep it simple with stage name, since I expect to remove the need for this coverage definition as we will eventually add generic coverage opt-in for both live and ci pipelines.

Comment (JianpingChen) [reply eng/pipelines/templates/stages/archetype-sdk-tests.yml:117]: Honestly, I think archetype-sdk-tests-generate.yml may not be the best place since each sdk may still overwrite UnsupportedClouds in its own tests.yml (with archetype-sdk-tests.yml as template) unless you move the top level iteration in archetype-sdk-tests.yml to higher level. To me, the closest place to overwrite the cloud value (regardless which variable holds the value) is right within the iteration over the clouds (currently in archetype-sdk-tests.yml).

Comment (JianpingChen) [reply eng/pipelines/templates/stages/archetype-sdk-tests.yml:117]: > This value should already be available via `$(System.StageName)`.

I can give that a try

Comment (JianpingChen) [reply eng/pipelines/templates/stages/archetype-sdk-tests.yml:117]: Actually, I don't see $(System.StageName) being set in archetype-sdk-tests.yml. If it take value from the "stage" name, more than likely it'll read like "Public_LiveTest"

**Comment by JianpingChen** [comment]
/azp check-enforcer override

Servicebus has test failures. Not my fault. :-)

File: sdk/communication/azure-communication-chat/tests.yml [hunk 2]
```diff
@@ -19,5 +19,13 @@ stages:
         groupId: com.azure
         safeName: azurecommunicationchat
     ServiceDirectory: communication
+    PostSteps:
+      - task: PublishCodeCoverageResults@1
```
Comment (weshaggard) [inline sdk/communication/azure-communication-chat/tests.yml:23]: We should try to avoid duplicating these steps everywhere. In fact we probably should avoid duplicating a set of these steps so by creating a template you can share across your packages.

File: eng/pipelines/templates/stages/archetype-sdk-tests.yml (hunk 1 above)
Comment (benbp) [reply eng/pipelines/templates/stages/archetype-sdk-tests.yml:117]: Ah, yep didn't catch that Java was appending some strings to the stage name. You could try the more flexible `contains($(System.StageName), 'Public')` or `startsWith(...)` if you didn't want to hardcode the whole thing.

Comment (benbp) [reply eng/pipelines/templates/stages/archetype-sdk-tests.yml:117]: I wasn't suggesting to define the cloud config, etc. in the generate file, but rather to set the variable there, e.g.

```
pwsh: |
  Write-Host "##vso[task.setvariable variable=CloudName;]${{ parameters.CloudConfig.Cloud }}"
```

Comment (benbp) [reply eng/pipelines/templates/stages/archetype-sdk-tests.yml:117]: All this said, I think similar to the comments in the .net PR, we should expect to make this more generic and not have to pass in these commands from the tests.yml file.

File: sdk/communication/azure-communication-chat/tests.yml (hunk 2 above)
Comment (JianpingChen) [reply sdk/communication/azure-communication-chat/tests.yml:23]: Let me try

Comment (JianpingChen) [reply sdk/communication/azure-communication-chat/tests.yml:23]: Made a few attempts. Sorry, they didn't work out. It is not obvious to me how this can be done without modifying code in eng\pipelines folder. If adding a new variable can be a concern, I am afraid I shouldn't be making big changes there for the purpose of this particular PR. @weshaggard

Comment (weshaggard) [reply sdk/communication/azure-communication-chat/tests.yml:23]: I've not tested these but this is what I was thinking:

Each individual template would look like
```
trigger: none

stages:
- template: /sdk/communication/tests-template.yml
  parameters:
    PackageName: azure-communication-chat
    SafeName: azurecommunicationchat
```
Add a new tests-template.yml file in the communication directory that looks something like:
```parameters:
  PackageName: not-specified
  SafeName: notspecified

stages:
- template: /eng/pipelines/templates/stages/archetype-sdk-tests.yml
  parameters:
    CloudConfig:
      Public:
        SubscriptionConfigurations:
          - $(sub-config-azure-cloud-test-resources)
          - $(sub-config-communication-services-cloud-test-resources-common)
          - $(sub-config-communication-services-cloud-test-resources-java)
      Int:
        SubscriptionConfigurations:
          - $(sub-config-communication-int-test-resources-common)
          - $(sub-config-communication-int-test-resources-java)
    Clouds: Public, Int
    Artifacts:
      - name: ${{ parameters.PackageName }}
        groupId: com.azure
        safeName: ${{ parameters.SafeName }}
    ServiceDirectory: communication
    PostSteps:
      - task: PublishCodeCoverageResults@1
        condition: and(eq(variables['Agent.OS'], 'Windows_NT'), eq(variables['JavaTestVersion'], '1.11'), eq(variables['CurrentCloud'], 'Public'))
        inputs:
          codeCoverageTool: JaCoCo
          summaryFileLocation: sdk/communication/${{ parameters.PackageName }}/target/site/jacoco/jacoco.xml
          reportDirectory: sdk/communication/${{ parameters.PackageName }}/target/site/jacoco/
          failIfCoverageEmpty: true
    EnvVars:
      SKIP_LIVE_TEST: TRUE
```

That would eliminate a lot of duplication and make sure things remain consistent across packages.

Comment (JianpingChen) [reply sdk/communication/azure-communication-chat/tests.yml:23]: I think I tried that and it complained about parameters declaration. Let me try again.

Comment (JianpingChen) [reply sdk/communication/azure-communication-chat/tests.yml]: @weshaggard That worked! Thank you! Updated. Please take another look

File: eng/pipelines/templates/stages/archetype-sdk-tests.yml (hunk 1 above)
Comment (JianpingChen) [reply eng/pipelines/templates/stages/archetype-sdk-tests.yml:117]: I think generalization is a good idea. But modifying engineering system, especially code under eng\pipelines, is not our expertise. It can be done after this works for us. :-)

Comment (weshaggard) [reply eng/pipelines/templates/stages/archetype-sdk-tests.yml:117]: While I'm not necessarily opposed to adding a new parameter here I do think we old do what @benbp suggested for you current usage (i.e. `startsWith($(System.StageName), 'Public')`)

File: sdk/communication/communication-tests-template.yml [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+            - $(sub-config-communication-int-test-resources-java)
+      Clouds: Public, Int
+      Artifacts:
+        - name: ${{ parameters.PackageName }}
+          groupId: com.azure
+          safeName: ${{ parameters.SafeName }}
+      ServiceDirectory: communication
+      PostSteps:
+        - task: PublishCodeCoverageResults@1
+          condition: and(eq(variables['Agent.OS'], 'Windows_NT'), eq(variables['JavaTestVersion'], '1.11'), eq(variables['CurrentCloud'], 'Public'))
```
Comment (weshaggard) [inline sdk/communication/communication-tests-template.yml:26]:
```suggestion
          condition: and(eq(variables['Agent.OS'], 'Windows_NT'), eq(variables['JavaTestVersion'], '1.11'), startsWith(variables['System.StageName'], 'Public'))
```

File: sdk/communication/azure-communication-chat/tests.yml (hunk 2 above)
Comment (weshaggard) [reply sdk/communication/azure-communication-chat/tests.yml]: Yes that looks like what I was looking for and helps reduce lot of duplication and increases consistency.

**Review by benbp — APPROVED** [review]
